demo/src/main/scala/starwars/StarWarsData.scala

```diff
@@ -0,0 +1,182 @@
+    name: Option[String],
+    appearsIn: Option[List[Episode.Value]],
+    primaryFunction: Option[String]
+  ) (friends0: => Option[List[Character]]) extends Character {
+    lazy val friends = friends0
+  }
+  object Droid {
+    def apply(
+      id: String,
+      name: Option[String],
+      appearsIn: Option[List[Episode.Value]],
+      friends: => Option[List[Character]],
+      primaryFunction: Option[String]
+    ) = new Droid(id, name, appearsIn, primaryFunction)(friends)
+  }
+  // #model_types
+
+  // #model_values
+  lazy val LukeSkywalker: Character =
+    Human(
+      id = "1000",
+      name = Some("Luke Skywalker"),
+      friends = Some(List(HanSolo, LeiaOrgana, C3PO, R2D2)),
+      appearsIn = Some(List(Episode.NEWHOPE, Episode.EMPIRE, Episode.JEDI)),
+      homePlanet = Some("Tatooine"))
+  lazy val DarthVader: Character =
+    Human(
+      id = "1001",
+      name = Some("Darth Vader"),
+      friends = Some(List(WilhuffTarkin)),
+      appearsIn = Some(List(Episode.NEWHOPE, Episode.EMPIRE, Episode.JEDI)),
+      homePlanet = Some("Tatooine"))
+  lazy val HanSolo: Character =
+    Human(
+      id = "1002",
+      name = Some("Han Solo"),
+      friends = Some(List(LukeSkywalker, LeiaOrgana, R2D2)),
+      appearsIn = Some(List(Episode.NEWHOPE, Episode.EMPIRE, Episode.JEDI)),
+      homePlanet = None)
+  lazy val LeiaOrgana: Character =
+    Human(
+      id = "1003",
+      name = Some("Leia Organa"),
+      friends = Some(List(LukeSkywalker, HanSolo, C3PO, R2D2)),
+      appearsIn = Some(List(Episode.NEWHOPE, Episode.EMPIRE, Episode.JEDI)),
+      homePlanet = Some("Alderaan"))
+  lazy val WilhuffTarkin: Character =
+    Human(
+      id = "1004",
+      name = Some("Wilhuff Tarkin"),
+      friends = Some(List(DarthVader)),
+      appearsIn = Some(List(Episode.NEWHOPE, Episode.EMPIRE, Episode.JEDI)),
+      homePlanet = None)
+  lazy val C3PO: Character =
+    Droid(
+      id = "2000",
+      name = Some("C-3PO"),
+      friends = Some(List(LukeSkywalker, HanSolo, LeiaOrgana, R2D2)),
+      appearsIn = Some(List(Episode.NEWHOPE, Episode.EMPIRE, Episode.JEDI)),
+      primaryFunction = Some("Protocol"))
+  lazy val R2D2: Character =
+    Droid(
+      id = "2001",
+      name = Some("R2-D2"),
+      friends = Some(List(LukeSkywalker, HanSolo, LeiaOrgana)),
+      appearsIn = Some(List(Episode.NEWHOPE, Episode.EMPIRE, Episode.JEDI)),
+      primaryFunction = Some("Astromech"))
+
+  // The character database ...
+  val characters = List(
+    LukeSkywalker, DarthVader, HanSolo, LeiaOrgana, WilhuffTarkin, C3PO, R2D2
+  )
+  // #model_values
+
+  import Episode._
+
+  // Mapping from Episode to its hero Character
+  lazy val hero: Map[Value, Character] = Map(
+    NEWHOPE -> R2D2,
+    EMPIRE -> LukeSkywalker,
+    JEDI -> R2D2
+  )
+}
+
+object StarWarsQueryCompiler extends QueryCompiler(StarWarsSchema) {
+  // #elaborator
+  val selectElaborator = new SelectElaborator(Map(
+    StarWarsSchema.tpe("Query").dealias -> {
```

Comment:
Yes, `tpe` does always return a `TypeRef`. For the most part you shouldn't need to know whether you have a `TypeRef` or the underlying non-`TypeRef`. The main situation where you do is when comparing types for equivalance, because `==` distinguishes `TypeRef`s from their underlying non-`TypeRef`. The `dealias` you can see here is a consequence of that: if you want to compare by `==` (and `hashCode`) you need to dealias on both side. I actually want to tweak the API a little bit here so that it's something a user has to do in this context, but that's for a separate PR.

Yes, if you add to the schema, but don't update the elaborator, then this will break at runtime. At some point it would make sense to have these elaborators generated or derived from the schema and the Scala data type, which would reduce that burden, or eliminate it altogether in most cases.